stdlib/public/core/String.swift

```diff
@@ -405,15 +405,31 @@ extension String {
   // This force type-casts element to UInt8, since we cannot currently
   // communicate to the type checker that we proved this with our dynamic
   // check in String(decoding:as:).
+  // **Important:** The 'repairsMade' part of the result is not reliable.
   @_alwaysEmitIntoClient
   @inline(never) // slow-path
   private static func _fromNonContiguousUnsafeBitcastUTF8Repairing<
     C: Collection
   >(_ input: C) -> (result: String, repairsMade: Bool) {
     _internalInvariant(C.Element.self == UInt8.self)
-    return Array(input).withUnsafeBufferPointer {
-      let raw = UnsafeRawBufferPointer($0)
-      return String._fromUTF8Repairing(raw.bindMemory(to: UInt8.self))
+    if #available(macOS 10.16, iOS 14.0, watchOS 7.0, tvOS 14.0, *) {
+      let count = input.count
+      let string = String(unsafeUninitializedCapacity: count) { buffer in
+        var iterAndCount = unsafeBitCast(
+          buffer, to: UnsafeMutableBufferPointer<C.Element>.self
+        ).initialize(from: input)
+        _precondition(
+          iterAndCount.0.next() == nil,
+          "Collection contains more than 'count' elements"
+        )
+        return iterAndCount.1
+      }
+      return (result: string, repairsMade: false)
+    } else {
+      return Array(input).withUnsafeBufferPointer {
+        let raw = UnsafeRawBufferPointer($0)
+        return String._fromUTF8Repairing(raw.bindMemory(to: UInt8.self))
+      }
```

Comment:
Is this the approach we want, or would we prefer to use `withContiguousStorageIfAvailable` on the input?

Reply:
This is in `_fromNonContiguousUnsafeBitcastUTF8Repairing`, so it's in the non-contiguous fall-back path.

The array copying here is the current implementation, and was done because I think we were doing UTF-8 validation on the contents in their source collection before copying them over, and that was just way faster on contiguous storage (the benefits were greater than the cost of the initial copy).

Anyway, now we have `String(unsafeUninitializedCapacity:...)`, which works on a different model where it first allocates storage, initialises the data, then validates and repairs its own contiguous storage. That means the source collection is only relevant for the initial copy.

Reply:
This function is called after the input has been dynamically established to have no contiguous storage, so this is fine. It would be better to use `withMemoryRebound` here or, more cheaply, `assumingMemoryBound`.

Reply:
I see nothing wrong with the following for the fallback:
`$0.withMemoryRebound(to: UInt8.self, String._fromUTF8Repairing)`
Or we could wait for `assumingMemoryBound` for the raw buffer.